aten/src/ATen/native/cuda/BatchLinearAlgebra.cu

```diff
@@ -1263,6 +1289,94 @@ std::tuple<Tensor, Tensor> _symeig_helper_cuda(const Tensor& self, bool eigenvec
+#else
+  magma_vec_t jobvr = eigenvectors ? MagmaVec : MagmaNoVec;
+  magma_int_t n = magma_int_cast(self.size(-1), "n");
+  auto self_data = self.data_ptr<scalar_t>();
+
+  out_eigvals.resize_({2, n});
+  auto out_eigvals_data = out_eigvals.data_ptr<scalar_t>();
+  scalar_t *wr = out_eigvals_data;
+  scalar_t *wi = out_eigvals_data+n;
+
+  scalar_t *vr_data = NULL;
+  magma_int_t ldvr = 1;
+  if (jobvr == MagmaVec)
+  {
+      vr_data = out_eigvecs.data_ptr<scalar_t>();
+      ldvr = n;
+  }
+
+  if (n > 0) {
+    // call magmaEig once to get the optimal size of work_data
+    scalar_t wkopt;
+    magmaEig<scalar_t>(MagmaNoVec, jobvr, n, self_data, n, wr, wi, NULL, 1, vr_data, ldvr, &wkopt, -1, info_ptr);
+    magma_int_t lwork = (magma_int_t) wkopt;
+
+    // call it a 2nd time to to the actual work
+    scalar_t *work_data = nullptr;
+    ALLOCATE_ARRAY(work_data, scalar_t, lwork);
+    magmaEig<scalar_t>(MagmaNoVec, jobvr, n, self_data, n, wr, wi, NULL, 1, vr_data, ldvr, work_data, lwork, info_ptr);
+  }
+
+  // magmaEig returns column-ordered tensors, turn them into row-ordered
+  out_eigvals.transpose_(0, 1);
+  out_eigvecs.transpose_(0, 1);
+#endif
+}
+
+
+std::tuple<Tensor &,Tensor &> eig_cuda_out(Tensor & e, Tensor & v, const Tensor & self, bool eigenvectors) {
+  Tensor vals_tmp, vecs_tmp;
+  std::tie(vals_tmp, vecs_tmp) = eig_cuda(self, eigenvectors);
+  e.resize_as_(vals_tmp).copy_(vals_tmp);
+  v.resize_as_(vecs_tmp).copy_(vecs_tmp);
+  return std::tuple<Tensor&, Tensor&>(e, v);
+}
+
+std::tuple<Tensor,Tensor> eig_cuda(const Tensor & self, bool eigenvectors) {
+  squareCheckInputs(self);
+  int64_t n = self.size(-1);
+
+  if (self.numel() == 0) {
+    auto eigvals = at::empty({n, 2}, self.options());
+    return std::tuple<Tensor, Tensor>(eigvals, at::empty_like(self, LEGACY_CONTIGUOUS_MEMORY_FORMAT));
+  }
+
+  // copy self to pinned CPU memory
+  auto self_working_copy = at::empty(self.sizes(),
+                                     at::TensorOptions(at::kCPU).dtype(self.dtype()).pinned_memory(true));
+
+  self_working_copy.transpose_(0, 1); // make it column-ordered, what magmaEig wants
+  self_working_copy.copy_(self);
+
+  // tensors holding the results
+  auto out_eigvals = at::empty(0, self.options().device(at::kCPU), LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+  auto out_eigvecs = eigenvectors
+                     ? at::empty({n, n}, self.options().device(at::kCPU), LEGACY_CONTIGUOUS_MEMORY_FORMAT)
+                     : Tensor();
+
+  magma_int_t info;
+  AT_DISPATCH_FLOATING_TYPES(self.scalar_type(), "eig_cuda", [&]{
+    apply_eig<scalar_t>(self_working_copy, eigenvectors, out_eigvals, out_eigvecs, &info);
+  });
+  singleCheckErrors(info, "eig_cuda");
+
+  if (eigenvectors) {
+      return std::tuple<Tensor, Tensor>(out_eigvals.to(self.device()), out_eigvecs.to(self.device()));
+  } else {
+      return std::tuple<Tensor, Tensor>(out_eigvals.to(self.device()), at::empty({0}, self.options()));
+  }
```

Comment:
1. copying from GPU to CPU is [what the old code does](https://github.com/pytorch/pytorch/blob/master/aten/src/THC/generic/THCTensorMathMagma.cu#L175): shouldn't I just follow the same logic, and possibly add the optimization in another PR?

2. is this a correct optimization, anyway? I assumed that the output Tensor should be on the same device as the input one, isn't it the case?

Reply:
Note that the MAGMA driver routines are hybrid algorithms and potentially have a somewhat complex interplay between CPU and GPU. Here the comment in the SymEig port might be likely to apply to Eig as well:
https://github.com/pytorch/pytorch/blob/9a5a7328660d4c205cb2ec2c3db341fb00ea8d3f/aten/src/ATen/native/cuda/BatchLinearAlgebra.cu#L1236-L1243

More generally: Should we expect the code to be very parallel between symeig and eig? To the point that we might share things with templates?

@pearu What is your take on what is the required device here?

Reply:
AFAIK, magma does not have `_gpu` variant of `geev` like `syev` has. Hence we cannot make `symeig` and `eig` code as alike as one would wish.

> * copying from GPU to CPU is [what the old code does](https://github.com/pytorch/pytorch/blob/master/aten/src/THC/generic/THCTensorMathMagma.cu#L175): shouldn't I just follow the same logic, and possibly add the optimization in another PR?

Let me copy from above: While the THC implementation is a good reference for porting, if there is a way to improve the existing code with respect to efficiency/simplicity, I think one should use the opportunity, IMHO.

> * is this a correct optimization, anyway? I assumed that the output Tensor should be on the same device as the input one, isn't it the case?

Assume that the input device is GPU. The answer depends on what users want to do with the output Tensor: (i) continue with the computations on GPU or (ii) access the output data from CPU. For the case (ii), copying the output to GPU will be unnecessary as for accessing the data, it must be copied back to CPU. For the case (i), the routines using the output from eig likely handle the copy of CPU data to GPU automatically anyway. Hence, keeping the output data where it was created seems more optimal to me than insisting that the input and output data must have the same locations, especially because in either way users do not need to change their code.

Reply:
You are right about geev needing to move the input to CPU, but this:
> For the case (ii), copying the output to GPU will be unnecessary as for accessing the data, it must be copied back to CPU.

Are you very sure about that? You are advising people to introduce BC breaking changes here. I must admit I have doubts about it.

Reply:
@t-vi , thanks for pointing out the BC breakage issue.

Reply:
Sorry to interject here, but have we considered reusing PyTorch's CPU dgeev (in TH) here? That is, have we actually compared the performance of GPU -> CPU -> TH dgeev -> GPU vs. this MAGMA-based approach? 

As for the out argument, by the way, PyTorch is moving to a model where we will not resize given outputs (unless they have no storage, in which case we interpret it as the user requesting a resize). However, resizing outputs or, in some cases, having to copy to them is currently deprecated, and it may more appropriate to throw a warning if that behavior was previously supported.

Reply:
yes, marking this conversation resolved.
Benchmarking `GPU -> CPU -> TH dgeev -> GPU vs. this MAGMA-based` would be valuable, imho.